lms/services/ltia_http.py

```diff
@@ -0,0 +1,72 @@
+import uuid
+from datetime import datetime, timedelta
+
+from lms.services import JWTService
+
+
+class LTIAHTTPService:
+    """Send LTI Advantage requests and return the responses."""
+
+    def __init__(self, lti_registration, jwt_service, http):
+        self._lti_registration = lti_registration
+        self._jwt_service = jwt_service
+        self._http = http
+
+    def _sign(self, registration, payload, lifetime=timedelta(hours=1)):
+        now = datetime.utcnow()
+        default_payload = {
+            "exp": now + lifetime,
+            "iat": now,
+            "nonce": uuid.uuid4().hex,
+            "iss": registration.client_id,
+            "sub": registration.client_id,
+        }
+        payload = dict(default_payload, **payload)
+
+        return self._jwt_service.encode_with_private_key(payload)
+
+    def _get_access_token(self, scopes):
+        """
+        Get an access token from the LMS to use in LTA services.
+
+        https://datatracker.ietf.org/doc/html/rfc7523
+        https://canvas.instructure.com/doc/api/file.oauth_endpoints.html#post-login-oauth2-token
+        """
+        jwt = self._sign(
+            self._lti_registration,
+            {
+                "jti": uuid.uuid4().hex,
+                "aud": self._lti_registration.token_url,
+            },
+        )
+        auth_request = {
+            "grant_type": "client_credentials",
+            "client_assertion_type": "urn:ietf:params:oauth:client-assertion-type:jwt-bearer",
+            "client_assertion": jwt,
+            "scope": " ".join(scopes),
+        }
+        response = self._http.post(
+            self._lti_registration.token_url,
+            data=auth_request,
+        )
+        return response.json()["access_token"]
+
+    def request(self, scopes, method, url, headers=None, **kwargs):
```

Comment:
```suggestion
    def request(self, method, url, headers=None, scopes=None, **kwargs):
        if not scopes:
            raise TypeError("Scopes are required")
```

I realise the defaulting here is awkward, but I feel like this quite strongly breaks the interface you'd expect by having scopes first.

If we did this it would also nicely mix with the suggestion to sub-class above.

Reply:
I agree it doesn't make much sense as the first param, move it after `url`. 

Not a fan of doing the `scopes=None` and then having to check against it being none. I guess with my proposal you could still pass None as the scope but the same is true of method and url.